src/libraries/Common/src/SourceGenerators/ImmutableEquatableArray.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics.Hashing;
+
+namespace SourceGenerators
+{
+    /// <summary>
+    /// Provides an immutable list implementation which implements sequence equality.
+    /// </summary>
+    public sealed class ImmutableEquatableArray<T> : IEquatable<ImmutableEquatableArray<T>>, IReadOnlyList<T>
+        where T : IEquatable<T>
+    {
+        public static ImmutableEquatableArray<T> Empty { get; } = new ImmutableEquatableArray<T>(Array.Empty<T>());
+
+        private readonly T[] _values;
+        public T this[int index] => _values[index];
```

Comment:
```suggestion
        public ref T this[int index] => ref _values[index];
```

This provides a more array like semantic

Reply:
But it also makes it mutable?

Reply:
If the goal is immutable then make the return `readonly ref`. 

If the goal is immutalbe though then why are we inroducing `record` with `set` members?

Reply:
I think we should aim for the model being immutable everywhere.

Reply:
FWIW it seems `readonly ref` isn't supported on indexers yet.

Reply:
Sorry, should be `ref readonly` not `readonly ref`